Support arbitrary blob signing (#283)
notaryproject/notaryproject: specs/signing-and-verification-workflow.md

# Signing and Verification Workflow

> This document describes the workflow of signing and verifying artifacts likes OCI images and arbitrary blobs.

Reviewer: ```suggestion
This document describes the workflow for signing and verifying OCI artifacts and arbitrary blobs.
```

            1. Verify that the timestamp signing algorithm satisfies [algorithm requirements](./signature-specification.md#signature-algorithm-requirements).
        1. Embed timestamp to the signature envelope.
1. **Save the signature envelope:** Save the signature envelope generated in the previous step to a file. File extension should be the original blob file name plus `.sig.jws` for JWS signatures and `.sig.cose` for COSE signatures.

## Arbitrary blob verification workflow

The user wants to consume an arbitrary blob only if it was signed by a trusted publisher and the signature associated with the blob is valid.

### Verification Prerequisites

- User has the blob that they want to consume, along with its detached signature.
- User has configured [trust store and trust policy](./trust-store-trust-policy.md) required for signature verification.

### Verification Steps

1. **Should implementations of this specification verify the signature? :** Depending upon [trust-policy](./trust-store-trust-policy.md#blob-trust-policy) configuration, determine whether implementations of this specification need to verify the signature or not.

>    If signature verification should be skipped for the given blob, skip the below steps and directly jump to step 4.

Reviewer 2: The "step 4" seems not correct. Should it skip all the rest steps or skip to step 3?

Author: Good catch. If the verification is skipped, then notation has no additional steps. Since the user already has the blob and they can choose to consume it (or not). In the OCI verification workflow, the "step 4" is downloading the OCI image.

I have updated the language accordingly.

           If all signature artifact descriptors have already been processed, fail the signature verification and exit.
1. **Get OCI artifact:** Using the verified digest, download the OCI artifact.
   This step is not in the purview of Notary Project.

> ## Arbitrary blob signing workflow

Reviewer 3: Please use consistent terminology. For e.g. use blob (not arbitrary blob) and blob signature(not detached signature)
```suggestion
## Blob signing workflow
```

The user wants to sign an arbitrary blob with a detached signature.

### Signing Prerequisites

- User has access to the signing certificate and private key or a remote signing service through a notation plug-in.

### Signing Steps

1. **Generate signature:** Using notation CLI or any other compliant signing tool, sign the blob. The signing tool should follow the following guideline.
    1. Construct the blob payload as defined in [`signature specification`](./signature-specification.md#payload)
    1. Verify that the signing certificate is valid and satisfies [certificate requirements](./signature-specification.md#certificate-requirements).
    1. Verify that the signing algorithm satisfies [algorithm requirements](./signature-specification.md#signature-algorithm-requirements).
    1. Generate signature.

>         1. Generate signature using signature formats specified in [supported signature envelopes](./signature-specification.md#supported-signature-envelopes). Also, as part of this step, the user-defined/supplied custom attributes should be added to the annotations of the signature's descriptor.

Reviewer: ```suggestion
        1. Generate signature using a signature format specified in [supported signature envelopes](./signature-specification.md#supported-signature-envelopes). Also, as part of this step, the user-defined/supplied custom attributes should be added to the annotations of the signature payload.
```

        1. If the user wants to timestamp the signature, obtain an [RFC-3161](https://datatracker.ietf.org/doc/html/rfc3161.html) compliant timestamp for the signature generated in the previous step. Otherwise, continue to the next step.
            1. Verify that the timestamp signing certificate satisfies [certificate requirements](./signature-specification.md#certificate-requirements).
            1. Verify that the timestamp signing algorithm satisfies [algorithm requirements](./signature-specification.md#signature-algorithm-requirements).
        1. Embed timestamp to the signature envelope.

> 1. **Save the signature envelope:** Save the signature envelope generated in the previous step to a file.

Reviewer: Should we add a line about the file extensions to use for each envelope type?

> 1. **Save the signature envelope:** Save the signature envelope generated in the previous step to a file. File extension should be the original blob file name plus `.sig.jws` for JWS signatures and `.sig.cose` for COSE signatures.

Reviewer 3: Should we use jws.sig for the file extension? 
Many application launchers determine file type based on characters after the last period. This aligns with your suggestion for the trust policy file extension as well.

Author: .sig is actually not a popular extension for digital signatures, unlike json. Different tools use different extensions based on their requirements. I think for our usecase, it is okay to go with either .sig.jws or .jws.sig. Both seem fine to me. Unless we have a strong reason to flip the order, lets stick to what we have.

Reviewer 3: I more inclined towards consistency with trust policy file name unless we have a usecase/requirement that doesn't works with `.sig`. Also .sig for signature is a popular extension used by many tools such as cosign, [in email signature](https://removefile.com/sig-file-extension/), etc

This shouldn't block this significant PR. If you decide to merge this PR, please open an issue to discuss this further

Author: Consistency with trust policies seems like a weak argument since json files are meant to be read and understood by various tools while signature files are meant for Notation only (since only Notation can understand the signature format we have). Again, I'm fine either ways. I don't have a strong opinion here. @shizhMSFT do you have a preference here?

Reviewer 2: TBH, I'd prefer `.sig` for all JWS and COSE signatures (and use other mechanisms like PEM or file magic numbers) but I'm fine that we distinguish them using file extension `.sig.jws` and `.sig.cose`.

In the case we are going the way of `sig.jws` and `sig.cose`, I'd prefer as it is. Taking `.tar.gz` as an example, `foo.tar.gz` is a gzip file containing the content of `foo.tar`. Hence, a JWS/COSE file file containing the content of a Notary Project signature is better to have the extension name of `.sig.jws` / `.sig.cose` instead of `.jws.sig` and `.cose.sig`.

In the case we are going the way of using `.sig` for all JWS and COSE signatures eventually, I'd like to say that `.jws.sig` and `.cose.sig` are better since both `.jws.sig` and `.cose.sig` are `.sig` files.

Note: email signatures use `.p7s` (pkcs#7/CMS) for the signature extension.

Author: Typically speaking, users think the token after the last dot as the file extension and prefer not to modify it. In that sense, keeping .sig.jws and .sig.cose yields better results than .jws.sig or .cose.sig (here, I'm referring to notation not needing to guess the format by trying all possibilities as a better result)

## Arbitrary blob verification workflow

The user wants to consume an arbitrary blob only if it was signed by a trusted publisher and the signature associated with the blob is valid.

### Verification Prerequisites

- User has the blob that they want to consume, along with its detached signature.
- User has configured [trust store and trust policy](./trust-store-trust-policy.md) required for signature verification.

### Verification Steps

> 1. **Should implementations of this specification verify the signature? :** Depending upon [trust-policy](./trust-store-trust-policy.md#trust-policy) configuration, determine whether implementations of this specification need to verify the signature or not.

Reviewer 3: Thinking more about this, should we even support skipping signature verification for blobs? 
The usecase of blobs is difference from registries were user might want to use single trust policy to verify there whole application (running multiple images/sidecars) .

Author: I was wondering about the same. If a user want to skip verification for a particular class of blobs, they can simply skip calling the notation verify command. I think one situation where supporting "skip" for blob verification makes sense is admission controllers where admins want to make notation a "mandatory step" in their CI/CD pipeline and let notation decide whether to skip verification or not on a particular blob. 

@gokarnm @shizhMSFT Any thoughts on this?

Reviewer 2: Having a `skip` options for the entire verification is like to support `none` alg in JWT verification. It is technically doable but I don't think it is a good security practice. It is strange to observe an artifact has passed the verification without any verification.

Author: > It is strange to observe an artifact has passed the verification without any verification.

Notation doesn't say that it passed the verification. It says that, "for the given conditions, notation is skipping the verification". Notation is already supporting "skip" for OCI images. @shizhMSFT are you suggesting that we should remove "skip" altogether?

Reviewer 2: > Notation doesn't say that it passed the verification. It says that, "for the given conditions, notation is skipping the verification".

What's the return value of the process in that case? Think about that `notation` will be used in scripting.

Reviewer 3: I think we can omit skip for blobs until we find a usecase and if required we can always add it later

Author: Looking at the `notation verify` command [documentation](https://github.com/notaryproject/notation/blob/main/specs/commandline/verify.md), It doesn't seem like it supports any contracts for return codes. In other words, there are no granular return codes for various verification outcomes. That being said, supporting "skip" option is not going to complicate any of existing notation contracts. I think we should *try* to keep "skip" experience consistent across different usecases.

Reviewer 2: Based on 
https://github.com/notaryproject/specifications/blob/f52bc09736a444d5dc340adb684d12c1bd78aeb7/specs/trust-store-trust-policy.md?plain=1#L295-L301
It seems we can support `skip`.

   If signature verification should be skipped for the given blob, skip the below steps and directly jump to step 4.

> 1. **verify the detached signature:**

Reviewer 3: ```suggestion
1. **Verify the detached signature:**
```

    1. Parse and validate the signature envelope using the detached signature's file extension as the envelope type.
    1. Verify the signature envelope using trust-store and trust-policy as mentioned in [signature evaluation](./trust-store-trust-policy.md#signature-evaluation) section.
    1. If the signature verification fails, exit.

> 1. Construct the blob payload as defined in [`signature specification`](./signature-specification.md#payload)

Reviewer 2: Since the payload is a JSON object, how do you make sure that the newly constructed blob payload is exactly the same as the payload constructed by the signer?

How to canonicalize the JSON object? For instance, `{"foo":1,"bar":2}` and `{"bar":2,"foo":1}` are identical JSON objects but different strings.

How do verifiers deal with user metadata?

Reviewer 2: It does not seem right. Should this step be something like "extract the blob payload from the signature envelope"?

Author: > Since the payload is a JSON object, how do you make sure that the newly constructed blob payload is exactly the same as the payload constructed by the signer?

> How to canonicalize the JSON object? For instance, {"foo":1,"bar":2} and {"bar":2,"foo":1} are identical JSON objects but different strings.

Notation is not verifying the string equality of the original payload vs the signed payload (in Json). The workflow here is calling out the verification on Size, MediaType and the digest. Which is what is being done for OCI artifacts as well. See https://github.com/oras-project/oras-go/blob/be5973602aa86b007b948be27ddc9c5749580b34/content/descriptor.go#L38

> It does not seem right. Should this step be something like "extract the blob payload from the signature envelope"?

Take a look at [this code](https://github.com/notaryproject/notation-go/blob/main/verifier/verifier.go#L173). In the case of OCI, original payload is already available (from the registry) and we are comparing it against the payload retrieved from the signature envelope. However, that is not the case for blobs. Hence, we need to construct the original payload from the blob itself and then compare it against `targetArtifact.payload`.

> How do verifiers deal with user metadata?

Good call out. Will add it.

Author: I modified the language a bit to avoid any confusion. Let me know if you still have any concerns.

> 1. Compare the payload derived from the above step with the payload present in the signature envelope. Fail signature verification if there is mismatch.

Reviewer: Maybe combine this with the next step. The constructed payload does not contain annotations, so this is bit confusing.

Author: Call out individual attributes that we are matching.

> 1. Verify blob's media type and size match the values present in `targetArtifact.payload`. Fail signature verification if there is a mismatch.

Reviewer 3: Should this be part of `Parse and validate the signature envelope using the detached signature's file extension as the envelope type.` step? Integrity validation is must and should be done before expiry and revocation check.

> 1. Calculate the digest of the blob using the digest algorithm specified at `targetArtifact.payload.digest` and make sure the blob digest matches the digest present in `targetArtifact.payload.digest`. Fail signature verification if there is a mismatch.

Reviewer 3: We should limit parsing of blob to the size specified in signature.

> 1. Calculate the digest of the blob using the digest algorithm specified at `targetArtifact.payload.digest` and make sure the blob digest matches the digest present in the `targetArtifact.payload.digest`. If the digests are equal, signature verification is considered successful

Reviewer: What about comparing media type and size?

> 1. Calculate the digest of the blob using the digest algorithm specified at `targetArtifact.payload.digest` and make sure the blob digest matches the digest present in `targetArtifact.payload.digest`. If they match, signature verification is considered successful.

Reviewer 2: If `digest algorithm` MUST be deduced from signing certificate's public key, should the verification step also verify whether the digest algorithm matches the verification key?

Author: Updated the language.

> 1. Verify blob's media type and size match the values present in `targetArtifact.payload`. If they match, signature verification is considered successful.

Reviewer 2: Should we verify the media type and size first so that we can fail fast?

Author: Sure

Reviewer 2: > Sure

Then let's change the verification order so that implementors can follow.
